File: exercise/java/day24/src/main/java/domain/core/RegisteredRoutes.java
package domain.core;

import io.vavr.collection.Map;
import io.vavr.collection.HashMap;

import java.util.function.Consumer;

public class RegisteredRoutes {
    private Map<String, Consumer<Event>> routes = HashMap.empty();

    public void dispatch(Event event) {
        routes.get(event.getClass().getName()).get().accept(event);
    }

    public void register(String eventType, Consumer<Event> apply) {
        routes = routes.put(eventType, apply);
    }
}

File: exercise/scala/day24/src/main/scala/domain/core/RegisteredRoutes.scala
package domain.core

import java.util.function.Consumer
import scala.collection.mutable

class RegisteredRoutes {
  private val routes: mutable.Map[String, Consumer[Event]] = mutable.Map.empty

  def dispatch(event: Event): Unit = routes.get(event.getClass.getName).foreach(_.accept(event))

  def register(eventType: String, apply: Consumer[Event]): Unit = routes += (eventType -> apply)
}